File: logic/levels.py
from logic.levelCreator import calculateScore

levels = {
    "1": {
        "costs": {
            "red": 1,
            "blue": 1,
            "green": 2,
            "orange": 5,
            "purple": lambda x: 2 * x,
            "yellow": lambda y: 3 * y,
        },
        "woodToCollect": 50,
        "waterToCollect": 20,
        "trees": 20,
        "rateOfWater": 5,
        "budget": 300,
    },
    "2": {
        "costs": {
            "red": 5,
            "blue": 10,
            "green": 2,
            "orange": 10,
            "purple": lambda x: 30 * x * x,
            "yellow": lambda y: 20 * y,
        },
        "woodToCollect": 80,
        "waterToCollect": 20,
        "trees": 5,
        "rateOfWater": 10,
        "budget": 500,
    },
}


def levelSelector(level: int, blobDetails: list) -> int:
    levelDetails = levels[str(level)]
    if not levelDetails:
        raise Exception("Invalid level")
    return max(
        0,
        calculateScore(
            blobDetails=blobDetails,
            woodToCollect=levelDetails["woodToCollect"],
            waterToCollect=levelDetails["waterToCollect"],
            trees=levelDetails["trees"],
            rateOfWater=levelDetails["rateOfWater"],
            costs=levelDetails["costs"],
            budget=levelDetails["budget"],
        ),
    )

File: game/src/models/Levels.js
import BlobCostCreator from "../models/BlobCostCreator";
import Environment from "../models/Environment";

function Levels(level) {
    const levels = {
        1: {
            // Creates the environment with (trees, waterFlow)
            env: new Environment(20, 5),
            // Sets the target resources to collect
            target: {
                wood: 50,
                water: 20,
            },
            costs: {
                // Red, blue, green and orange all have fixed costs
                red: new BlobCostCreator((_) => 1, "1"),
                blue: new BlobCostCreator((_) => 1, "1"),
                green: new BlobCostCreator((_) => 2, "2"),
                orange: new BlobCostCreator((_) => 5, "5"),
                // Purple and yellow have costs that depend on the number of repetitions
                purple: new BlobCostCreator(
                    (x) => 2 * x,
                    "2 * number of blobs repeated",
                ),
                yellow: new BlobCostCreator(
                    (x) => 3 * x,
                    "3 * number of repetitions",
                ),
            },
            // Sets the budget for the level. If the choice of blobs is successful, the
            // user will get budget-spent points. Here, 300 is used because you can
            // afford 50 orange blobs, and 20 blue blos with some points left over.
            budget: 300,
        },
        2: {
            env: new Environment(5, 10),
            target: {
                wood: 80,
                water: 20,
            },
            costs: {
                red: new BlobCostCreator((_) => 1, "5"),
                blue: new BlobCostCreator((_) => 1, "10"),
                green: new BlobCostCreator((_) => 2, "2"),
                orange: new BlobCostCreator((_) => 5, "10"),
                purple: new BlobCostCreator(
                    (x) => 3 * x * x,
                    "30 * (number of blobs repeated)^2",
                ),
                yellow: new BlobCostCreator(
                    (x) => 20 * x,
                    "20 * number of repetitions",
                ),
            },
            budget: 500,
        },
    };

    return levels[level];
}

export default Levels;
